### PR TITLE
NAS-123925 / 24.04 / Added JobProgressDialogComponent

### PR DESCRIPTION
**Pull request by RehanY147**
This creates a new way of showing job progress and subcribing to main job events i.e., Success, Abort, Failure and Progress updates.

### CODE REVIEW

File: src/app/modules/common/dialog/job-progress/job-progress-dialog.component.ts [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    onFailure?: (job: Job) => void;
+    onAbort?: (job: Job) => void;
+    onProgress: (progress: JobProgress) => void;
+  };
+  config: {
+    title: string;
+    description: string;
+  };
+  flags: {
+    showRealtimeLogs: boolean;
```
Comment (undsoft) [inline src/app/modules/common/dialog/job-progress/job-progress-dialog.component.ts:23]: realtime logs are only used in one form, I'd consider creating a separate component for this.

File: src/app/modules/common/dialog/job-progress/job-progress-dialog.component.ts [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    autoCloseOnSuccess: boolean;
```
Comment (undsoft) [inline src/app/modules/common/dialog/job-progress/job-progress-dialog.component.ts:24]: I think it's a good idea to decide this and have it work one way.

File: src/app/modules/common/dialog/job-progress/job-progress-dialog.component.ts [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+  };
+}
+
+@UntilDestroy()
+@Component({
+  templateUrl: './job-progress-dialog.component.html',
+  styleUrls: ['./job-progress-dialog.component.scss'],
```
Comment (undsoft) [inline src/app/modules/common/dialog/job-progress/job-progress-dialog.component.ts:31]: onpush

File: src/app/services/dialog.service.ts [hunk 4]
```diff
@@ -1,20 +1,85 @@
...
+import { Job, JobProgress } from 'app/interfaces/job.interface';
 import { ConfirmDialogComponent } from 'app/modules/common/dialog/confirm-dialog/confirm-dialog.component';
 import { ErrorDialogComponent } from 'app/modules/common/dialog/error-dialog/error-dialog.component';
 import { FullScreenDialogComponent } from 'app/modules/common/dialog/full-screen-dialog/full-screen-dialog.component';
 import { GeneralDialogComponent, GeneralDialogConfig } from 'app/modules/common/dialog/general-dialog/general-dialog.component';
 import { InfoDialogComponent } from 'app/modules/common/dialog/info-dialog/info-dialog.component';
+import { JobProgressDialogComponent, JobProgressDialogConfig } from 'app/modules/common/dialog/job-progress/job-progress-dialog.component';
 import { MultiErrorDialogComponent } from 'app/modules/common/dialog/multi-error-dialog/multi-error-dialog.component';
 
+export class JobProgressDialog {
```
Comment (undsoft) [inline src/app/services/dialog.service.ts:21]: Needs to go in some other file.

File: src/app/services/dialog.service.ts [hunk 5]
```diff
@@ -1,20 +1,85 @@
...
+  private readonly _onProgress$ = new Subject<JobProgress>();
+
+  readonly afterSuccess$ = this.onSuccess$.asObservable();
+  readonly afterAbort$ = this.onAbort$.asObservable();
+  readonly afterFailure$ = this.onFailure$.asObservable();
+  readonly onProgress$ = this._onProgress$.asObservable();
+
+  private _matDialogRef: MatDialogRef<JobProgressDialogComponent>;
+
+  set matDialogRef(ref: MatDialogRef<JobProgressDialogComponent>) {
```
Comment (undsoft) [inline src/app/services/dialog.service.ts:34]: Getter/setter doesn't seem to provide value.

File: src/app/services/dialog.service.ts [hunk 6]
```diff
@@ -1,20 +1,85 @@
...
 import { GeneralDialogComponent, GeneralDialogConfig } from 'app/modules/common/dialog/general-dialog/general-dialog.component';
 import { InfoDialogComponent } from 'app/modules/common/dialog/info-dialog/info-dialog.component';
+import { JobProgressDialogComponent, JobProgressDialogConfig } from 'app/modules/common/dialog/job-progress/job-progress-dialog.component';
 import { MultiErrorDialogComponent } from 'app/modules/common/dialog/multi-error-dialog/multi-error-dialog.component';
 
+export class JobProgressDialog {
+  private readonly onSuccess$ = new Subject<Job>();
+  private readonly onAbort$ = new Subject<Job>();
+  private readonly onFailure$ = new Subject<Job>();
+  private readonly _onProgress$ = new Subject<JobProgress>();
```
Comment (undsoft) [inline src/app/services/dialog.service.ts:25]: leading underscore doesn't match our styleguide

File: src/app/services/dialog.service.ts [hunk 7]
```diff
@@ -1,20 +1,85 @@
...
+
+  readonly afterSuccess$ = this.onSuccess$.asObservable();
+  readonly afterAbort$ = this.onAbort$.asObservable();
+  readonly afterFailure$ = this.onFailure$.asObservable();
+  readonly onProgress$ = this._onProgress$.asObservable();
```
Comment (undsoft) [inline src/app/services/dialog.service.ts:30]: What's the difference between using `asObservable` vs returning a subject, which is also an observable?

**Review by undsoft — CHANGES_REQUESTED** [review]
I think we need to decide whether we want to handle errors directly in jobProgress dialog.
Though it may be less customizable, I think it may be a good idea – most of the time you want to delegate everything about showing job status to this dialog.

File: src/app/services/dialog.service.ts (hunk 7 above)
Comment (RehanY147) [reply src/app/services/dialog.service.ts]: Subjects are allowed to emit values. The intention is to only expose buttons that are needed to the class's user. They don't need to do emits. So why expose that.

**Comment by RehanY147** [comment]
@undsoft For real time logs, I think we can get away with one flag instead of a whole component. Also, in regards to handling errors, sometimes the user might not want to do anything with an error. Such as when running data protection tasks on the Data Protection dashboard.

File: src/app/classes/job-progress-dialog-ref.class.ts [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+import { JobProgressDialogComponent } from 'app/modules/common/dialog/job-progress/job-progress-dialog.component';
+
+export class JobProgressDialogRef {
+  constructor(private readonly jobProgressDialog: JobProgressDialog) { }
+
+  getDialogRef(): MatDialogRef<JobProgressDialogComponent> {
+    return this.jobProgressDialog.matDialogRef;
+  }
+
+  onSuccess(): Observable<Job> {
```
Comment (undsoft) [inline src/app/classes/job-progress-dialog-ref.class.ts:14]: I think we may also need something called `onClose`.

File: src/app/services/dialog.service.ts [hunk 9]
```diff
@@ -120,4 +125,32 @@ export class DialogService {
...
     }
   }
+
+  jobProgress<M extends ApiJobMethod>(
+    { job$, title, description }: {
+      job$: Observable<Job<ApiJobDirectory[M]['response']>>;
+      title: string;
+      description: string;
+    },
+    showRealtimeLogs: boolean,
```
Comment (undsoft) [inline src/app/services/dialog.service.ts:135]: `showRealtimeLogs` then needs to be part of config object.

File: src/app/services/dialog.service.ts [hunk 10]
```diff
@@ -120,4 +125,32 @@ export class DialogService {
       openDialog.close();
     }
   }
+
+  jobProgress<M extends ApiJobMethod>(
+    { job$, title, description }: {
+      job$: Observable<Job<ApiJobDirectory[M]['response']>>;
+      title: string;
+      description: string;
```
Comment (undsoft) [inline src/app/services/dialog.service.ts:133]: Title and description should be optional.

**Review by undsoft — CHANGES_REQUESTED** [review]
I still think we should handle errors in job dialog. You as a developer delegate showing job status to a dialog when you open it. Error to me is also a type of a job status.
If you don't care about job failing, then to me you should use something like `startJob` without the dialog.
Currently if you mess with a job by altering call params, job dialog will keep showing progress indicator.

**Comment by RehanY147** [comment]
@undsoft Test on the team m40 and open the `DataProtectionDashboard` and a request is made the fails so far so you can see what happens when job fails.

File: src/app/classes/job-progress-dialog-ref.class.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+
+  onFailure(): Observable<Job> {
+    return this.jobProgressDialog.afterFailure$;
+  }
+
+  onProgressUpdate(): Observable<JobProgress> {
+    return this.jobProgressDialog.onProgress$;
+  }
+
+  onClose(): Observable<unknown> {
```
Comment (undsoft) [inline src/app/classes/job-progress-dialog-ref.class.ts:30]: `Observable<void>`

**Review by undsoft — APPROVED** [review]
I think it looks pretty good now.

**Pull request merged**